lib/Base/System/Session.vala

```diff
@@ -118,7 +118,7 @@ public abstract class Backend.Session : Object {
    *
    * @return The post created from the data.
    */
-  internal abstract Post load_post (Json.Object data);
+  internal abstract Post load_post (Json.Object data, bool force_load = false);
```

Comment:
If you make the new properties settable after construction, this is no longer needed, right?

Reply:
No, we still need it. If we make the new properties settable then we could just toggle the `is_favourited`/`is_reposted` flag _but_ Mastodon gives you an entire, object back that's normally been updated. So it'll include new reblog counts etc. So we don't have to modify the numbers ourselves.

This lets us bypass the caching so that we actually load that new object with the updated values. It may even be useful in other places - e.g. maybe we update the cache with the new object built from the new JSON when we search. Or when we load an account's timeline.

Reply:
I would argue that it makes more sense to update the properties of the existing object rather than to replace the whole object. Hence the `construct set` note, which admittedly also applies for metrics. I wanted to add an method that can update an post when it was cached later on. 

In this case, any new pull could just update the cached object, so an "force change" parameter wouldn't be needed.

Reply:
Just read this again. 

If I'm right, your point is that by recreating the object on change, we avoid to add an additional update method doing the same as the constructor, right?

My concern would be that creating and removing objects on the repost it would add more effort where the object is used in collections and such and also is probably not as good regarding memory usage.

Reply:
Yeah. "load from post" with a force flag lets us just create new objects from JSON without needing setters on everything. And it avoids duplicate code.

Updating the values on the instance would mean that every part of the app is holding. But there's no mechanism in place to to notify those changes, so what's shown will get out of sync.

For example: I view a thread. I reblog a reply. That `Post` object now has `is_reposted` set to True. I go back to the home timeline. The post is also visible in the home timeline.

If we created a new object then the home timeline still has the older `is_reposted = False` version. So clicking on the button will check the object state, decide it needs to reblog, make the HTTP request, and get an "already reblogged" response. At which point it will fetch the new version from the cache and show the new status.

If we update the object then the home timeline has the `is_reposted = True` status. BUT nothing notified it to update the UI. So the button looks like it's not reblogged. Then we click it and the logic says "the post is reblogged, therefore un-reblog it". We've got an unexpected action. The UI now shows the right state (unreblogged), but the user is confused because it looks like the click didn't do anything.

We could have notifications on properties that trigger the UI to refresh. But that feels like it's going to be _more_ overhead! And more potential for things remaining in memory (and causing crashes) because we didn't disconnect an event somewhere.

Reply:
For notifying the UI, we could use [`bind_property`](https://valadoc.org/gobject-2.0/GLib.Object.bind_property.html).

I should probably already have implemented it that way, would've made this discussion less complicated...

Reply:
Those bind properties are actually quite nice. If you ignore the fact that you need to specify them with strings. I didn't know about them before.

It seems to be working now with the latest commits.